mmv1/third_party/terraform/utils/framework_location_test.go

```diff
@@ -0,0 +1,235 @@
+}
+
+func TestLocationDescription_getRegion(t *testing.T) {
+	cases := map[string]struct {
+		ld             LocationDescription
+		ExpectedRegion types.String
+		ExpectedError  bool
+	}{
+		"returns the value of the region field in resource config": {
+			ld: LocationDescription{
+				ResourceRegion: types.StringValue("resource-region"),
+				ProviderRegion: types.StringValue("provider-region"),
+			},
+			ExpectedRegion: types.StringValue("resource-region"),
+		},
+		"shortens region values set as self links in the resource config": {
+			ld: LocationDescription{
+				ResourceRegion: types.StringValue("https://www.googleapis.com/compute/v1/projects/my-project/regions/us-central1"),
+			},
+			ExpectedRegion: types.StringValue("us-central1"),
+		},
+		"returns a region derived from the zone field in resource config when region is unset": {
+			ld: LocationDescription{
+				ResourceZone: types.StringValue("provider-zone-a"),
+			},
+			ExpectedRegion: types.StringValue("provider-zone"),
+		},
+		"does not shorten region values when derived from a zone self link set in the resource config": {
+			ld: LocationDescription{
+				ResourceZone: types.StringValue("https://www.googleapis.com/compute/v1/projects/my-project/zones/us-central1-a"),
+			},
+			ExpectedRegion: types.StringValue("https://www.googleapis.com/compute/v1/projects/my-project/zones/us-central1"), // Value isn't sortened from URI to name
+		},
+		"returns the value of the region field in provider config when region/zone is unset in resource config": {
+			ld: LocationDescription{
+				ProviderRegion: types.StringValue("provider-region"),
+			},
+			ExpectedRegion: types.StringValue("provider-region"),
+		},
+		"returns a region derived from the zone field in provider config when region unset in both resource and provider config": {
+			ld: LocationDescription{
+				ProviderZone: types.StringValue("provider-zone-a"),
+			},
+			ExpectedRegion: types.StringValue("provider-zone"),
+		},
+		"returns an error when zone values can't be found": {
+			ExpectedError: true,
+		},
+		"returns an error that mention non-standard schema field names when region value can't be found": {
+			ld: LocationDescription{
+				ZoneSchemaField: types.StringValue("foobar"),
+			},
+			ExpectedError: true,
+		},
+	}
+	for tn, tc := range cases {
+		t.Run(tn, func(t *testing.T) {
+
+			region, err := tc.ld.getRegion()
+
+			if err != nil {
+				if tc.ExpectedError {
+					if !tc.ld.RegionSchemaField.IsNull() {
+						if !strings.Contains(err.Error(), tc.ld.RegionSchemaField.ValueString()) {
+							t.Fatalf("expected error to use provider schema field value %s, instead got: %s", tc.ld.RegionSchemaField.ValueString(), err)
+						}
+					}
+					return
+				}
+				t.Fatalf("unexpected error using test: %s", err)
+			}
+			if region != tc.ExpectedRegion {
+				t.Fatalf("Incorrect region: got %s, want %s", region, tc.ExpectedRegion)
+			}
+		})
+	}
+}
+
+func TestLocationDescription_getLocation(t *testing.T) {
```

Comment:
Do we want any tests around if region and zone are set?

Reply:
There are test cases covering when region and zone are used to get a location value, but I've updated the first test case to show what happens when _all_ values are available (even though that wouldn't/shouldn't really happen in the provider) - this is added in [f8dd164](https://github.com/GoogleCloudPlatform/magic-modules/pull/7835/commits/f8dd164310dcccb122b3609c7dcceece3a1d6078)

Reply:
Could you also add a case where location is not available, but region and zone are, to make sure that region takes precedence if available? The current tests don't seem to cover that case.